### DISCUSSION TITLE
Is there a way to chain "derivative" instructions?

### DISCUSSION THREAD

**Discussion by ppravatto** [autodiff/autodiff#150] [Q&A]
I am new to the field of automatic differentiation and I was looking into it in order to migrate some of my codes from numerical differentiation (finite difference) to more accurate procedures. I often encounter multiple expressions with nested derivatives and I was wondering if there is a way to simply encode those using `autodiff`. I was thinking of something like the following pseudo-code:

```cpp
#include <iostream>
using namespace std;

#include <autodiff/forward.hpp>
using namespace autodiff;

using dual2nd = HigherOrderDual<2>; 

dual2nd f1(dual2nd x){
    return "some_function_of_x";
}

dual2nd f2(dual2nd x){
    return "some_function_of_x";
}

// Final function for which the first derivative is needed
dual2nd g(dual2nd x){
    return f2(x) * derivative(f1, wrt(x), at(x));
}

int main(){
    dual2nd x = 2.0;
    dual2nd u = g(x);

    double dudx = derivative(g, wrt(x), at(x));

    cout << "u = " << u << endl;
    cout << "du/dx = " << dudx << endl;
}
```

Is possible to do something like this or am I missing something fundamental? If this is possible can you point me to some example?

**Comment by allanleal** [comment]
I would suggest not to do this in function `g`:

~~~c++
f2(x) * derivative(f1, wrt(x), at(x));
~~~

because you are mixing 2nd order dual number (the return of `f2(x)`) and a 1st order dual number (the return of `derivative(f1, wrt(x), at(x))`). If you later decide to compute 2nd order derivatives of `g`, you will not have 2nd order derivative contributions coming from the branch `derivative(f1, wrt(x), at(x))`. 

Please do your experiments to see how things work.

**Reply by jusevitch** [reply]
@ppravatto Any updates on this? I'm interested in nested derivatives / gradients as well, but it's not clear how to do this. I haven't been able to get anything to work yet.

**Reply by allanleal** [reply]
Hi @jusevitch , could you please post an example of what you are trying to achieve? Maybe in a new question if not exactly the same as the one here.

**Reply by jusevitch** [reply]
Sure, I'll post a new question. Thanks for the response!

**Reply by jusevitch** [reply]
I created #255 with an example of what I have in mind.
